File: 프로그래머스/2/43165. 타겟 넘버/타겟 넘버.java
class Solution {
    public static int answer = 0;

    public int solution(int[] numbers, int target) {
        back(numbers, new int[numbers.length], 0, target);
        return Solution.answer;
    }

    // Back Tracking 구현
    public void back(int[] numbers, int[] arr, int cnt, int target) {
        // 기저 사례
        if (cnt == arr.length) {
            int sum = 0;
            for (int i = 0; i < arr.length; i++) {
                sum += arr[i];
            }
            if (sum == target) Solution.answer += 1;
            return;
        }

        arr[cnt] = numbers[cnt];
        back(numbers, arr, cnt + 1, target);
        arr[cnt] = -numbers[cnt];
        back(numbers, arr, cnt + 1, target);
    }
}

File: 프로그래머스/2/43165. 타겟 넘버/타겟 넘버.py
answer = 0

def dfs(numbers, target, idx, current_sum):
    global answer
    if idx == len(numbers):
        if current_sum == target:
            answer += 1
        return
    
    current_num = numbers[idx]
    dfs(numbers, target, idx + 1, current_sum + current_num)
    dfs(numbers, target, idx + 1, current_sum - current_num)
    

def solution(numbers, target):
    global answer
    dfs(numbers, target, 0, 0)
    return answer